### PR TITLE
feat: add deduplication ratio to 'ipfs dag stat'

### LINKED ISSUE #9762: `ipfs dag stat` should accept multiple CIDs and show deduplication ratio
It is a commun ask to calculate deduplication ratio of many CIDs.

This would be easy to do if done like this:
```go
$ ipfs dag stat Qmfoo Qmbar
Size: 21465226420, NumBlocks: 85369, SizeDedup: 1.5, NumBlocksDedup: 1.01
```

### PR DESCRIPTION
**Pull request by arthurgavazza**
This adds deduplication information on a list of Cids for the `ipfs dag stat` command,as requested in https://github.com/ipfs/kubo/issues/9762. ex: 
running `ipfs dag stat QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU` writes the following output to stdout: 

```
CID                                           	Blocks         	Size
QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27

Summary

Total Size: 54
Unique Blocks: 2
Shared Size: 27
Ratio: 1.500000
```

and with the `--enc=json` option: 

```
{
    "UniqueBlocks": 2,
    "TotalSize": 54,
    "SharedSize": 27,
    "Ratio": 1.5,
    "DagStats": [
        {
            "Cid": "QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE",
            "Size": 27,
            "NumBlocks": 1
        },
        {
            "Cid": "QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU",
            "Size": 27,
            "NumBlocks": 1
        },
        {
            "Cid": "QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU",
            "Size": 27,
            "NumBlocks": 1
        }
    ]
}


```

Fixes #9762

### CODE REVIEW

**Comment by arthurgavazza** [comment]
@Jorropo this is just a draft, I'm still working on it, I'll add the tests and improve some things before starting with the bonus points. Can you take a look at this, just to see if I'm in the right direction? I don't know if that's what you wanted for the json and text encoders.

**Comment by arthurgavazza** [comment]
@Jorropo can you review this draft? I'm kind of blocked on this feature, I just need to know if I'm on the right direction and if this satisfies the issue basic requirements so I can continue working on it and make the improvements you mentioned.

**Comment by BigLep** [comment]
Just wanted to say thanks @arthurgavazza for your contribution.  The maintainers have been busy with https://2023.ipfs-thing.io/ but will be back into more normal flow next week.

**Comment by arthurgavazza** [comment]
@BigLep thanks for letting me know!

File: core/commands/dag/dag.go
Comment (Jorropo) [inline core/commands/dag/dag.go:319]: I think this needs comment explaining why we can't rely on `cid.Cid.MarshalJSON` and `cid.Cid.UnmarshalJSON` (since they do `{"/": "..."}` and it's not what the Kubo API do and we would rather be consistent).

File: core/commands/dag/dag.go
Comment (Jorropo) [inline core/commands/dag/dag.go:347]: I think `calculateRatio` and `calculateSharedSize` should be merged in one method that just computes everything, I don't think you would care about one of them without caring about the other.

File: core/commands/dag/dag.go
Comment (Jorropo) [inline core/commands/dag/dag.go:379]:
```suggestion
			if err := csvWriter.Write(header); err != nil {
			  return err
			}
```

File: core/commands/dag/stat.go
Comment (Jorropo) [inline core/commands/dag/stat.go:63]: AFAIT you fully removed support for the `progressive` option, we want to keep that, even if the information is partial (such as a simple total count and size sum equivalent to the previous option).

File: core/commands/dag/stat.go
Comment (Jorropo) [inline core/commands/dag/stat.go:11]:
```suggestion
	"github.com/ipfs/boxo/ipld/merkledag/traverse"
```

File: core/commands/dag/dag.go
Comment (Jorropo) [inline core/commands/dag/dag.go:392]:
```suggestion
			fmt.Print("\nSummary\n")
```

**Review by Jorropo — CHANGES_REQUESTED** [review]
Thx a lot <3 this is good.
I have a few comments.

File: core/commands/dag/dag.go
Comment (Jorropo) [inline core/commands/dag/dag.go:375]: Oh and I forgot, but all the writes here need to use `fmt.F*` functions, and pass `w` in the first argument.
```suggestion
			fmt.Fprintln(w)
```

**Review by Jorropo — CHANGES_REQUESTED** [review]

**Comment by arthurgavazza** [comment]
@Jorropo thx for the feedback! I'll work on this ASAP

**Comment by Jorropo** [comment]
For ci fix please run:
```bash
find -type f -name "go.mod" | while read f; do cd $(dirname "${f}"); go mod tidy; cd -; done
go fmt ./...
```

File: core/commands/dag/stat.go
Comment (arthurgavazza) [reply core/commands/dag/stat.go:63]: I'm thinking of a way to output the progressive info for each CID and make it nicely formatted. Could it be like this? 

```
CID                                           	Blocks         	Size
QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE	2200000        	20
QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE	2200000        	20
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2200000        	20
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2200000        	20


Final Stats

CID                                           	Blocks         	Size
QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27

Summary

Total Size: 54
Unique Blocks: 2
Shared Size: 27
Ratio: 1.500000
```

I would make a tsv table with the progressive data and once it's done, Print a line indicating that it's the final stats and print the final table with the summary.

Comment (Jorropo) [reply core/commands/dag/stat.go:63]: Whatever you want, the current solution is extremely simple.

I basically want a way for users to see "something is happening". A simple counter that shows total blocks downloaded (like we have now) is fine.
If you want something more fancy feel free to do it but don't feel obligated to. :slightly_smiling_face:

Comment (arthurgavazza) [reply core/commands/dag/stat.go:63]: I'd rather focus on the other improvements you mentioned than doing something fancy for this right now. Can I keep the current solution with the same format for now?  It would probably look like this

```
Size: 21465226420, NumBlocks: 85379
Size: 21465226430, NumBlocks: 85389
Size: 21465226440, NumBlocks: 85399

Final Stats

CID                                           	Blocks         	Size
QmcQs9AFNh4mHKqFLJ2Zo7WPSottWi9YDdEoqdk3jv5RyE	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27
QmSg4hRo947Lkp9qKn3KJhjjCFGNmtw8VLoXhUCC9mLRgU	2500000        	27

Summary

Total Size: 54
Unique Blocks: 2
Shared Size: 27
Ratio: 1.500000
```

Comment (Jorropo) [reply core/commands/dag/stat.go:63]: Look good to me.

**Comment by BigLep** [comment]
@arthurgavazza: are you to fix CI?  It would be great to include in the next release.

**Comment by Jorropo** [comment]
I think some fixes after the review havn't been pushed yet to. :slightly_smiling_face:

**Comment by arthurgavazza** [comment]
@BigLep @Jorropo I've been busy with work this week. I still have to finish some fixes, I'm gonna try to do this until tomorrow night if that's ok

**Comment by Jorropo** [comment]
@arthurgavazza you don't have any deadline, feel free to work on this when you have time. It will get in the release cycle whenever that happens.
Thx for your time this is appreciated.

File: test/cli/dag_test.go
Comment (arthurgavazza) [inline test/cli/dag_test.go:1]: @Jorropo I finished some of the fixes and I'm gonna start coding the tests. I think that for testing this functionality I'm gonna need a few test files, so that I can work with objects that have multiple blocks. Should I add a test files folder in test/cli ?

Comment (Jorropo) [reply test/cli/dag_test.go:1]: We keep tests in the `test/cli` folder, you can helpers elsewhere.
I'm surprised you would need more than one file, A golden test is good enough
1. manually create a dag that expose the deduping behaviour we want and store it as a `.car` file in the repo
2. in the test `ipfs dag import` it.
3. Compare the result of `ipfs dag stat ...` vs a set of known correct values from the car file above (called a golden file).

Comment (arthurgavazza) [reply test/cli/dag_test.go:1]: thanks for the tips!

Comment (arthurgavazza) [inline test/cli/dag_test.go:1]: @Jorropo I added a test that validates the output of the command using a dag that I've created, Is that ok?
Should I also add tests for the json encoding output and the progressive information?
Also, The sharness tests for this command are failing, since the output format changed, should I remove those?

File: test/cli/dag_test.go
Comment (Jorropo) [inline test/cli/dag_test.go:36]: Please use the \` syntax, this is very hard to me.

File: test/cli/dag_test.go
Comment (Jorropo) [inline test/cli/dag_test.go:33]:
```suggestion
		assert.NoError(t, err)
```

File: test/cli/dag_test.go
Comment (Jorropo) [reply test/cli/dag_test.go:1]: Look good to me thx.
Feel free to remove the sharness tests if they are also covered in the new CLI (which looks like they are).

**Review by Jorropo — CHANGES_REQUESTED** [review]

File: test/cli/dag_test.go
Comment (arthurgavazza) [reply test/cli/dag_test.go:36]: sorry for that, I'm gonna make this more readable

Comment (arthurgavazza) [reply test/cli/dag_test.go:36]: @Jorropo I'm having a hard time comparing the text output with a string literal in this test, I added a test using the json encoded output so that I can validate the deduplication behaviour. Is there any other way that I could test the text output?

Comment (Jorropo) [reply test/cli/dag_test.go]: @arthurgavazza json is good.

For the text one: the point of golden tests is that the code of the test is very dumb.
You create a dataset ahead of time that show all the behaviour you want (here it would be a `.car` fixture) and then you `==` or `diff` the output with some hardcoded values you create at the same time you make your fixture.

**Comment by arthurgavazza** [comment]
@Jorropo sorry for taking too long to make the changes, I was very busy this last few weeks. I added a new test, comparing the text output with a fixture containing the expected results. Can you  please take a look at the latest changes?

File: core/commands/dag/dag.go
Comment (hacdias) [inline core/commands/dag/dag.go:280]:
```suggestion
```

**Review by hacdias — COMMENTED** [review]
Very happy that you also moved the old relevant tests from `.sh` to `.go` 😃

File: test/cli/dag_test.go
Comment (Jorropo) [inline test/cli/dag_test.go:100]:
```suggestion
		content, err := io.ReadAll(f)
```
`f.Read` is allowed to partially fill the buffer even tho the OS might have all the data, various oses do this differently, for example if you have a very slow disk the OS will likely trickle data slowly to you.
All of this to say the `n` value from `.Read` is critical as you might do shorter reads than you expect, the simplest solution to solve this is to put this code in a loop and keep reading, it's what `io.ReadAll` do.

**Review by Jorropo — CHANGES_REQUESTED** [review]
Thx <3

**Comment by Jorropo** [comment]
Thx <3 :tada:

**Comment by arthurgavazza** [comment]
@Jorropo thx for all the support! ❤️  I'm really glad I could help. About the improvements you mentioned on the issue. Can I open a new issue with them and start working on it?

**Comment by Jorropo** [comment]
@arthurgavazza which improvements ?

**Comment by arthurgavazza** [comment]
@Jorropo 
Bonus points (you can go for it if you want to have fun but I don't care about it for now):

You execute the multiple traversals in parallel of one an other. That means all iterations of the loop starts a goroutine and you wg.Wait() the final result before printing it.
You add computation of internal compression ratios too (that means CIDs that share blocks with themselves multiple times).
You optimize the code by caching results of the traversal at any point to only traverse blocks once, even if they are duplicated across the different CIDs (hard).
You compute a result in a single pass (medium if done alone, very hard if combined with 1 and 3).
You output rich progress information similar to the final output (instead of just some total amount blocks).

**Pull request merged**
